Warning  from Clang static analyser

Hi,

Using the GROMACS CI for the MDModule version, the job of Clang static analyser failed (https://gitlab.com/gromacs/gromacs/-/jobs/4668733430) and reported some warnings in the colvars code which we can fix or ignore 

```
Found compiler warning during build
../../../src/external/colvars/colvaratoms.cpp:549:3: warning: Value stored to 'parse_error' is never read [deadcode.DeadStores]
  parse_error |= parse_fitting_options(group_conf);
  ^              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
../../../src/external/colvars/colvar.cpp:1884:57: warning: Although the value stored to 'delta' is used in the enclosing expression, the value is never actually read from 'delta' [deadcode.DeadStores]
    if ((is_enabled(f_cv_reflecting_lower_boundary) && (delta = x_ext - lower_boundary) < 0) ||
                                                        ^       ~~~~~~~~~~~~~~~~~~~~~~
../../../src/external/colvars/colvar.cpp:1885:57: warning: Although the value stored to 'delta' is used in the enclosing expression, the value is never actually read from 'delta' [deadcode.DeadStores]
        (is_enabled(f_cv_reflecting_upper_boundary) && (delta = x_ext - upper_boundary) > 0)) {
                                                        ^       ~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
../../../src/external/colvars/colvarcomp_angles.cpp:352:5: warning: Value stored to 'rA' is never read [deadcode.DeadStores]
    rA = 1.0;
    ^    ~~~
../../../src/external/colvars/colvarcomp_angles.cpp:366:5: warning: Value stored to 'rC' is never read [deadcode.DeadStores]
    rC = 1.0;
    ^    ~~~
2 warnings generated.
../../../src/external/colvars/colvarbias_meta.cpp:1603:19: warning: 2nd function call argument is an uninitialized value [core.CallAndMessage]
  hills.push_back(hill(h_it, h_weight, h_centers, h_sigmas, h_replica));
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/external/colvars/colvarbias_meta.cpp:1642:14: warning: Potential leak of memory pointed to by 'pwd' [cplusplus.NewDeleteLeaks]
      return cvm::error("Error: cannot get the path of the current working directory.\n",
             ^~~~~~~~~~
2 warnings generated.
../../../src/external/colvars/colvarparse.cpp:788:13: warning: Value stored to 'data_end' is never read [deadcode.DeadStores]
            data_end = brace+1;
            ^          ~~~~~~~
../../../src/external/colvars/colvarparse.cpp:795:11: warning: Value stored to 'data_end' is never read [deadcode.DeadStores]
          data_end = brace+1;
          ^          ~~~~~~~
2 warnings generated.
../../../src/external/colvars/colvarcomp_gpath.cpp:52:13: warning: 5th function call argument is an uninitialized value [core.CallAndMessage]
            cvm::load_coords(reference_position_filename.c_str(), &reference_position, atoms, reference_column, reference_column_value);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/external/colvars/colvarcomp_gpath.cpp:83:13: warning: 5th function call argument is an uninitialized value [core.CallAndMessage]
            cvm::load_coords(reference_position_filename.c_str(), &reference_fitting_position, tmp_fitting_atoms, reference_column, reference_column_value);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
../../../src/external/colvars/colvarmodule.cpp:971:7: warning: Value stored to 'error_code' is never read [deadcode.DeadStores]
      error_code |= proxy->smp_biases_script_loop();
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../src/external/colvars/colvarmodule.cpp:974:7: warning: Value stored to 'error_code' is never read [deadcode.DeadStores]
      error_code |= proxy->smp_biases_loop();
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

--- a/.github/workflows/backend-template.yml
+++ b/.github/workflows/backend-template.yml
@@ -62,6 +62,8 @@ on:
 jobs:
   test_backends:
     runs-on: ubuntu-latest
+    env:
+      CCACHE_DIR: ${{ github.workspace }}/ccache_${{ inputs.container_name }}
 
     steps:
       - name: Checkout Colvars
@@ -74,8 +76,7 @@ jobs:
         uses: actions/cache@v3
         with:
           path: |
-            ~/.ccache
-            ~/ccache_${{ inputs.container_name }}
+            ${{ github.workspace }}/ccache_${{ inputs.container_name }}
           key: ${{ runner.os }}-build-${{ inputs.backend_name }}-${{ github.sha }}
           restore-keys: |
             ${{ runner.os }}-build-${{ inputs.backend_name }}-
@@ -145,7 +146,6 @@ jobs:
           CMAKE_GENERATOR: Ninja
         run: |
           apptainer exec devel-tools/${{ inputs.container_name }}.sif ./update-colvars-code.sh -f ${{ inputs.backend_name }}-source
-          CCACHE_DIR=~/ccache_${{ inputs.container_name }} \
           apptainer exec devel-tools/${{ inputs.container_name }}.sif \
           bash ${{ inputs.path_compile_script }} ${{ inputs.backend_name }}-source
 

--- a/.github/workflows/test-library.yml
+++ b/.github/workflows/test-library.yml
@@ -159,6 +159,54 @@ jobs:
         uses: github/codeql-action/analyze@v2
 
 
+  static-analyser:
+    name: Clang 14 Static Analyser
+    runs-on: ubuntu-latest
+    needs: basicchecks
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Install build dependencies for library
+        run: |
+          sudo apt -y install tcl8.6-dev clang-14 clang-tools-14
+
+      - name: Checkout OpenMM (for Lepton library)
+        uses: actions/checkout@v3
+        with:
+          repository: 'openmm/openmm'
+          ref: 'master'
+          path: 'openmm-source'
+
+      - name: Perform Clang Static Analyser
+        env:
+          CXX: clang++-14
+          CC: clang-14
+        run: |
+          scan-build-14 --use-cc=clang-14 --use-c++=clang++-14 cmake -P devel-tools/build_test_library.cmake 2>&1 | tee build_analyser.log
+          grep "CMake Error" build_analyser.log > build_errors.log || true
+          if [ -s build_errors.log ] ; then
+            echo "Found compiler errors during build"
+            cat build_analyser.log
+            exit 1
+          fi
+
+      - name: Analyze logs
+        run: | # Retrieve warnings generated but not the ones of OpenMM
+          awk '/colvars\/colvars\/[^openmm].*warning:/,/warning.*generated|^$/' build_analyser.log | tee warnings_analyser.log || true
+          if [ -s warnings_analyser.log ] ; then
+            echo "Found compiler warning during build"
+            exit 1
+          fi
+
+      - name: Archive warnings artifacts
+        uses: actions/upload-artifact@v3
+        if: failure()
+        with:
+          name: warnings_clang_static_analyser
+          path: warnings_analyser.log
+
+
   build-linux-x86_64-many:
     name: Linux x86_64 (many compilers)
     runs-on: ubuntu-latest

--- a/.github/workflows/test-library.yml
+++ b/.github/workflows/test-library.yml
@@ -208,7 +208,7 @@ jobs:
 
 
   build-linux-x86_64-many:
-    name: Linux x86_64 (many compilers)
+    name: Linux x86_64 (GCC, Clang)
     runs-on: ubuntu-latest
     needs: basicchecks
     if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository

--- a/.github/workflows/test-library.yml
+++ b/.github/workflows/test-library.yml
@@ -12,6 +12,9 @@ jobs:
     if: github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository
     outputs:
       hassecrets: ${{ steps.checksecrets.outputs.hassecrets }}
+    env:
+      CCACHE: ccache
+      CCACHE_DIR: ${{ github.workspace }}/ccache
 
     steps:
       - name: Checkout Colvars
@@ -20,7 +23,7 @@ jobs:
       - name: Load compiler cache
         uses: actions/cache@v3
         with:
-          path: ~/.ccache
+          path: ${{ github.workspace }}/ccache
           key: ${{ runner.os }}-build-basic-${{ github.sha }}
           restore-keys: ${{ runner.os }}-build-basic-
 
@@ -49,8 +52,6 @@ jobs:
           fi
 
       - name: Build and test library (GCC)
-        env:
-          CCACHE: ccache
         run: cmake -P devel-tools/build_test_library.cmake
 
       - name: Check documentation of command-line scripting interface
@@ -66,14 +67,11 @@ jobs:
       - name: Build and test library (Clang)
         env:
           CMAKE_BUILD_DIR: build-clang
-          CCACHE: ccache
           CXX: clang++
           CC: clang
         run: cmake -P devel-tools/build_test_library.cmake
 
       - name: Build library with debug flags enabled
-        env:
-          CCACHE: ccache
         run: cmake -D COLVARS_DEBUG=ON -P devel-tools/build_test_library.cmake
 
       - name: Check whether secrets are available
@@ -82,6 +80,9 @@ jobs:
           PULL_ORACLE_DEVELOPER_STUDIO: ${{ secrets.PULL_ORACLE_DEVELOPER_STUDIO }}
         run: |
           echo "hassecrets=${{ env.PULL_ORACLE_DEVELOPER_STUDIO != '' }}" >> $GITHUB_OUTPUT
+
+      - name: Print ccache stats
+        run: ccache -s -v
 
 
   build-doc:
@@ -215,6 +216,7 @@ jobs:
     env:
       CCACHE: ccache
       CMAKE_GENERATOR: Ninja
+      CCACHE_DIR: ${{ github.workspace }}/ccache
 
     steps:
       - uses: actions/checkout@v3
@@ -222,7 +224,7 @@ jobs:
       - name: Load compiler ccache
         uses: actions/cache@v3
         with:
-          path: ~/.ccache
+          path: ${{ github.workspace }}/ccache
           key: ${{ runner.os }}-build-multiple-${{ github.sha }}
           restore-keys: ${{ runner.os }}-build-multiple-
 

--- a/colvartools/poisson_integrator.cpp
+++ b/colvartools/poisson_integrator.cpp
@@ -6,13 +6,13 @@
 
 int main (int argc, char *argv[]) {
 
-  colvarproxy *proxy = new colvarproxy();
-  colvarmodule *colvars = new colvarmodule(proxy);
-
   if (argc < 2) {
     std::cerr << "One argument needed: file name.\n";
     return 1;
   }
+
+  colvarproxy *proxy = new colvarproxy();
+  colvarmodule *colvars = new colvarmodule(proxy);
 
   std::string gradfile (argv[1]);
   colvar_grid_gradient grad(gradfile);
@@ -28,5 +28,7 @@ int main (int argc, char *argv[]) {
 
   potential.write_multicol(std::string(gradfile + ".int"),
                            "integrated potential");
+
+  delete colvars;
   return 0;
 }

--- a/devel-tools/containers/CentOS7-devel.def
+++ b/devel-tools/containers/CentOS7-devel.def
@@ -13,10 +13,10 @@ From: centos:7
         redhat-lsb-core "@Development Tools" \
         ncurses which bc man-db vim emacs screen tmux \
         gcc gcc-c++ gcc-gfortran glibc-static libstdc++-static clang cppcheck \
-        autoconf automake  cvs git git-cvs cvsps subversion mercurial \
+        autoconf automake cvs git git-cvs cvsps subversion mercurial \
         rh-git227 devtoolset-{7..11} llvm-toolset-7 cmake3 ccache ninja-build \
         doxygen \
-        openmpi-devel fftw-devel tcl-devel \
+        Lmod openmpi-devel fftw-devel tcl-devel \
         python-{devel,virtualenv} numpy scipy tkinter \
         python3-{devel,tkinter} python36-{virtualenv,numpy,scipy} \
         rh-python38-python-{devel,tkinter,numpy,scipy} \

--- a/devel-tools/containers/CentOS9-devel.def
+++ b/devel-tools/containers/CentOS9-devel.def
@@ -19,10 +19,10 @@ From: quay.io/centos/centos:stream9
         bc man-db vim emacs tcsh \
         "@Development Tools" \
         gcc gcc-c++ gcc-gfortran cppcheck \
-        gcc-toolset-{12,13} \
+        gcc-toolset-{12,13} clang clang-tools-extra \
         autoconf automake cvs git subversion mercurial \
         cmake ccache ninja-build \
-        openmpi-devel tbb-devel fftw-devel tcl-devel \
+        Lmod openmpi-devel tbb-devel fftw-devel tcl-devel \
         python3-{devel,tkinter,virtualenv,numpy,scipy} \
         ncurses-devel \
         rlwrap \

--- a/src/colvar.cpp
+++ b/src/colvar.cpp
@@ -1883,8 +1883,8 @@ void colvar::update_extended_Lagrangian()
       (is_enabled(f_cv_reflecting_upper_boundary) && (delta = x_ext - upper_boundary) > 0)) {
     x_ext -= 2.0 * delta;
     v_ext *= -1.0;
-    if ((is_enabled(f_cv_reflecting_lower_boundary) && (delta = x_ext - lower_boundary) < 0) ||
-        (is_enabled(f_cv_reflecting_upper_boundary) && (delta = x_ext - upper_boundary) > 0)) {
+    if ((is_enabled(f_cv_reflecting_lower_boundary) && (x_ext - lower_boundary) < 0.0) ||
+        (is_enabled(f_cv_reflecting_upper_boundary) && (x_ext - upper_boundary) > 0.0)) {
       cvm::error("Error: extended coordinate value " + cvm::to_str(x_ext) + " is still outside boundaries after reflection.\n");
     }
   }

--- a/src/colvaratoms.cpp
+++ b/src/colvaratoms.cpp
@@ -600,7 +600,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
 
   if (is_enabled(f_ag_rotate)) setup_rotation_derivative();
 
-  return (cvm::get_error() ? COLVARS_ERROR : COLVARS_OK);
+  return parse_error;
 }
 
 

--- a/src/colvaratoms.cpp
+++ b/src/colvaratoms.cpp
@@ -398,7 +398,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
   // }
   // colvarparse::Parse_Mode mode = parse_normal;
 
-  int parse_error = COLVARS_OK;
+  int error_code = COLVARS_OK;
 
   // Optional group name will let other groups reuse atom definition
   if (get_keyval(group_conf, "name", name)) {
@@ -448,7 +448,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
         cvm::error("Error: cannot find atom group with name " + atoms_of + ".\n");
         return COLVARS_ERROR;
       }
-      parse_error |= add_atoms_of_group(ag);
+      error_code |= add_atoms_of_group(ag);
     }
   }
 
@@ -462,7 +462,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
     std::string numbers_conf = "";
     size_t pos = 0;
     while (key_lookup(group_conf, "atomNumbers", &numbers_conf, &pos)) {
-      parse_error |= add_atom_numbers(numbers_conf);
+      error_code |= add_atom_numbers(numbers_conf);
       numbers_conf = "";
     }
   }
@@ -471,7 +471,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
     std::string index_group_name;
     if (get_keyval(group_conf, "indexGroup", index_group_name)) {
       // use an index group from the index file read globally
-      parse_error |= add_index_group(index_group_name);
+      error_code |= add_index_group(index_group_name);
     }
   }
 
@@ -480,7 +480,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
     size_t pos = 0;
     while (key_lookup(group_conf, "atomNumbersRange",
                       &range_conf, &pos)) {
-      parse_error |= add_atom_numbers_range(range_conf);
+      error_code |= add_atom_numbers_range(range_conf);
       range_conf = "";
     }
   }
@@ -507,7 +507,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
         cvm::error("Error: more instances of \"atomNameResidueRange\" than "
                    "values of \"psfSegID\".\n", COLVARS_INPUT_ERROR);
       } else {
-        parse_error |= add_atom_name_residue_range(psf_segids.size() ?
+        error_code |= add_atom_name_residue_range(psf_segids.size() ?
           *psii : std::string(""), range_conf);
         if (psf_segids.size()) psii++;
       }
@@ -533,25 +533,25 @@ int cvm::atom_group::parse(std::string const &group_conf)
       }
 
       // NOTE: calls to add_atom() and/or add_atom_id() are in the proxy-implemented function
-      parse_error |= cvm::load_atoms(atoms_file_name.c_str(), *this, atoms_col, atoms_col_value);
+      error_code |= cvm::load_atoms(atoms_file_name.c_str(), *this, atoms_col, atoms_col_value);
     }
   }
 
   // Catch any errors from all the initialization steps above
-  if (parse_error || cvm::get_error()) return (parse_error || cvm::get_error());
+  if (error_code || cvm::get_error()) return (error_code || cvm::get_error());
 
   // checks of doubly-counted atoms have been handled by add_atom() already
 
   if (get_keyval(group_conf, "dummyAtom", dummy_atom_pos, cvm::atom_pos())) {
 
-    parse_error |= set_dummy();
-    parse_error |= set_dummy_pos(dummy_atom_pos);
+    error_code |= set_dummy();
+    error_code |= set_dummy_pos(dummy_atom_pos);
 
   } else {
 
     if (!(atoms_ids.size())) {
-      parse_error |= cvm::error("Error: no atoms defined for atom group \""+
-                                key+"\".\n", COLVARS_INPUT_ERROR);
+      error_code |= cvm::error("Error: no atoms defined for atom group \"" + key + "\".\n",
+                               COLVARS_INPUT_ERROR);
     }
 
     // whether these atoms will ever receive forces or not
@@ -561,7 +561,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
   }
 
   // Now that atoms are defined we can parse the detailed fitting options
-  parse_error |= parse_fitting_options(group_conf);
+  error_code |= parse_fitting_options(group_conf);
 
   if (is_enabled(f_ag_scalable) && !b_dummy) {
     cvm::log("Enabling scalable calculation for group \""+this->key+"\".\n");
@@ -600,7 +600,7 @@ int cvm::atom_group::parse(std::string const &group_conf)
 
   if (is_enabled(f_ag_rotate)) setup_rotation_derivative();
 
-  return parse_error;
+  return error_code;
 }
 
 

--- a/src/colvarbias_meta.cpp
+++ b/src/colvarbias_meta.cpp
@@ -24,8 +24,8 @@
 #define PATHSEP "/"
 #endif
 
-// When C++17 is available, use the standard library's API
-#if (__cplusplus >= 201703L)
+#ifdef __cpp_lib_filesystem
+// When std::filesystem is available, use it
 #include <filesystem>
 #undef GETCWD
 #define GETCWD(BUF, SIZE) (std::filesystem::current_path().string().c_str())

--- a/src/colvarbias_meta.cpp
+++ b/src/colvarbias_meta.cpp
@@ -13,17 +13,22 @@
 #include <iomanip>
 #include <algorithm>
 
-// used to set the absolute path of a replica file
+// Define function to get the absolute path of a replica file
 #if defined(_WIN32) && !defined(__CYGWIN__)
 #include <direct.h>
-#define CHDIR ::_chdir
-#define GETCWD ::_getcwd
+#define GETCWD(BUF, SIZE) ::_getcwd(BUF, SIZE)
 #define PATHSEP "\\"
 #else
 #include <unistd.h>
-#define CHDIR ::chdir
-#define GETCWD ::getcwd
+#define GETCWD(BUF, SIZE) ::getcwd(BUF, SIZE)
 #define PATHSEP "/"
+#endif
+
+// When C++17 is available, use the standard library's API
+#if (__cplusplus >= 201703L)
+#include <filesystem>
+#undef GETCWD
+#define GETCWD(BUF, SIZE) (std::filesystem::current_path().string().c_str())
 #endif
 
 #include "colvarmodule.h"
@@ -1638,7 +1643,7 @@ int colvarbias_meta::setup_output()
 
     // TODO: one may want to specify the path manually for intricated filesystems?
     char *pwd = new char[3001];
-    if (GETCWD(pwd, 3000) == NULL) {
+    if (GETCWD(pwd, 3000) == nullptr) {
       return cvm::error("Error: cannot get the path of the current working directory.\n",
                         COLVARS_BUG_ERROR);
     }

--- a/src/colvarbias_meta.cpp
+++ b/src/colvarbias_meta.cpp
@@ -1644,6 +1644,9 @@ int colvarbias_meta::setup_output()
     // TODO: one may want to specify the path manually for intricated filesystems?
     char *pwd = new char[3001];
     if (GETCWD(pwd, 3000) == nullptr) {
+      if (pwd != nullptr) { //
+        delete[] pwd;
+      }
       return cvm::error("Error: cannot get the path of the current working directory.\n",
                         COLVARS_BUG_ERROR);
     }

--- a/src/colvarbias_meta.cpp
+++ b/src/colvarbias_meta.cpp
@@ -1537,7 +1537,7 @@ std::istream & colvarbias_meta::read_hill(std::istream &is)
   std::istringstream data_is(data);
 
   cvm::step_number h_it = 0L;
-  cvm::real h_weight;
+  cvm::real h_weight = 0.0;
   std::vector<colvarvalue> h_centers(num_variables());
   for (i = 0; i < num_variables(); i++) {
     h_centers[i].type(variables(i)->value());

--- a/src/colvarcomp_angles.cpp
+++ b/src/colvarcomp_angles.cpp
@@ -349,7 +349,7 @@ void colvar::dihedral::calc_gradients()
     A *= rA;
     cvm::rvector const dcosdA = rA*(cos_phi*A-B);
     cvm::rvector const dcosdB = rB*(cos_phi*B-A);
-    rA = 1.0;
+    // rA = 1.0;
 
     cvm::real const K = (1.0/sin_phi) * (180.0/PI);
 
@@ -363,7 +363,7 @@ void colvar::dihedral::calc_gradients()
     C *= rC;
     cvm::rvector const dsindC = rC*(sin_phi*C-B);
     cvm::rvector const dsindB = rB*(sin_phi*B-C);
-    rC = 1.0;
+    // rC = 1.0;
 
     cvm::real    const K = (-1.0/cos_phi) * (180.0/PI);
 

--- a/src/colvarcomp_gpath.cpp
+++ b/src/colvarcomp_gpath.cpp
@@ -31,9 +31,9 @@ colvar::CartesianBasedPath::CartesianBasedPath(std::string const &conf): cvc(con
     // Lookup reference column of PDB
     // Copied from the RMSD class
     std::string reference_column;
-    double reference_column_value;
+    double reference_column_value = 0.0;
     if (get_keyval(conf, "refPositionsCol", reference_column, std::string(""))) {
-        bool found = get_keyval(conf, "refPositionsColValue", reference_column_value, 0.0);
+        bool found = get_keyval(conf, "refPositionsColValue", reference_column_value, reference_column_value);
         if (found && reference_column_value == 0.0) {
           cvm::error("Error: refPositionsColValue, "
                      "if provided, must be non-zero.\n");

--- a/src/colvarmodule.cpp
+++ b/src/colvarmodule.cpp
@@ -994,7 +994,7 @@ int colvarmodule::calc_biases()
     total_bias_energy += (*bi)->get_energy();
   }
 
-  return (cvm::get_error() ? COLVARS_ERROR : COLVARS_OK);
+  return error_code;
 }
 
 

--- a/src/colvarmodule.cpp
+++ b/src/colvarmodule.cpp
@@ -327,6 +327,7 @@ void colvarmodule::config_changed()
 
 int colvarmodule::parse_global_params(std::string const &conf)
 {
+  int error_code = COLVARS_OK;
   // TODO document and then echo this keyword
   parse->get_keyval(conf, "logLevel", log_level_, log_level_,
                     colvarparse::parse_silent);
@@ -334,10 +335,7 @@ int colvarmodule::parse_global_params(std::string const &conf)
     std::string units;
     if (parse->get_keyval(conf, "units", units)) {
       units = colvarparse::to_lower_cppstr(units);
-      int error_code = proxy->set_unit_system(units, (colvars.size() != 0));
-      if (error_code != COLVARS_OK) {
-        return error_code;
-      }
+      error_code |= proxy->set_unit_system(units, (colvars.size() != 0));
     }
   }
 
@@ -346,7 +344,7 @@ int colvarmodule::parse_global_params(std::string const &conf)
     size_t pos = 0;
     while (parse->key_lookup(conf, "indexFile", &index_file_name, &pos)) {
       cvm::log("# indexFile = \""+index_file_name+"\"\n");
-      read_index_file(index_file_name.c_str());
+      error_code |= read_index_file(index_file_name.c_str());
       index_file_name.clear();
     }
   }
@@ -358,9 +356,8 @@ int colvarmodule::parse_global_params(std::string const &conf)
   }
 
   bool b_analysis = true;
-  if (parse->get_keyval(conf, "analysis", b_analysis, true,
-                        colvarparse::parse_silent)) {
-    cvm::log("Warning: keyword \"analysis\" is deprecated: it is now set "
+  if (parse->get_keyval(conf, "analysis", b_analysis, true, colvarparse::parse_silent)) {
+    cvm::log("Warning: keyword \"analysis\" is deprecated: it is now always set "
              "to true; individual analyses are performed only if requested.");
   }
 
@@ -391,7 +388,7 @@ int colvarmodule::parse_global_params(std::string const &conf)
   parse->get_keyval(conf, "sourceTclFile", source_Tcl_script);
 #endif
 
-  return cvm::get_error();
+  return error_code;
 }
 
 

--- a/src/colvarparse.cpp
+++ b/src/colvarparse.cpp
@@ -785,14 +785,12 @@ bool colvarparse::key_lookup(std::string const &conf,
           if (line[brace] == '{') brace_count++;
           if (line[brace] == '}') brace_count--;
           if (brace_count == 0) {
-            data_end = brace+1;
             break;
           }
           brace = line.find_first_of("{}", brace+1);
         }
 
         if (brace_count == 0) {
-          data_end = brace+1;
           break;
         }
 

